src/dev/run_check_lockfile_symlinks.js

```diff
@@ -0,0 +1,182 @@
+import { dirname } from 'path';
+
+import { run, createFailError } from '@kbn/dev-utils';
+
+import { REPO_ROOT } from './constants';
+import { File } from './file';
+import { matchesAnyGlob } from './globs';
+
+const LOCKFILE_GLOBS = ['**/yarn.lock'];
+const MANIFEST_GLOBS = ['**/package.json'];
+const IGNORE_FILE_GLOBS = [
+  // tests aren't used in production, ignore them
+  '**/test/**/*',
+  // fixtures aren't used in production, ignore them
+  '**/*fixtures*/**/*',
+  // cypress isn't used in production, ignore it
+  'x-pack/legacy/plugins/apm/cypress/*',
+];
+
+run(async ({ log }) => {
+  const paths = await globby(LOCKFILE_GLOBS.concat(MANIFEST_GLOBS), {
+    cwd: REPO_ROOT,
+    nodir: true,
+    gitignore: true,
+    ignore: [
+      // the gitignore: true option makes sure that we don't
+      // include files from node_modules in the result, but it still
+      // loads all of the files from node_modules before filtering
+      // so it's still super slow. This prevents loading the files
+      // and still relies on gitignore to to final ignores
+      '**/node_modules',
+    ],
+  });
+
+  const files = paths.map(path => new File(path));
+
+  await checkLockfileSymlinks(log, files);
+});
+
+async function checkLockfileSymlinks(log, files) {
+  const filtered = files.filter(file => !matchesAnyGlob(file.getRelativePath(), IGNORE_FILE_GLOBS));
+  await checkOnlyLockfileAtProjectRoot(filtered);
+  await checkSuperfluousSymlinks(log, filtered);
+  await checkMissingSymlinks(log, filtered);
+}
+
+async function checkOnlyLockfileAtProjectRoot(files) {
+  const errorPaths = [];
+
+  files
+    .filter(file => matchesAnyGlob(file.getRelativePath(), LOCKFILE_GLOBS))
+    .forEach(file => {
+      const path = file.getRelativePath();
+      const parent = dirname(path);
+      const stats = lstatSync(path);
+      if (!stats.isSymbolicLink() && parent !== '.') {
+        errorPaths.push(path);
+      }
+    });
+
+  if (errorPaths.length) {
+    throw createFailError(
+      `These directories MUST NOT have a 'yarn.lock' file:\n${listPaths(errorPaths)}`
+    );
+  }
+}
+
+async function checkSuperfluousSymlinks(log, files) {
+  const errorPaths = [];
+
+  files
+    .filter(file => matchesAnyGlob(file.getRelativePath(), LOCKFILE_GLOBS))
+    .forEach(file => {
+      const path = file.getRelativePath();
+      const parent = dirname(path);
+      const stats = lstatSync(path);
+      if (!stats.isSymbolicLink()) {
+        return;
+      }
+
+      const manifestPath = `${parent}/package.json`;
+      if (!existsSync(manifestPath)) {
+        log.warning(
+          `No manifest found at '${manifestPath}', but found an adjacent 'yarn.lock' symlink.`
+        );
+        errorPaths.push(path);
+        return;
+      }
+
+      try {
+        const manifest = readFileSync(manifestPath);
+        try {
+          const json = JSON.parse(manifest);
+          if (!json.dependencies || !Object.keys(json.dependencies).length) {
```

Comment:
Couldn't we have a scenario where there's `devDependencies`, but no `dependencies` and wouldn't that be ok?

Reply:
It would be OK, but since this PR is to support scanning tools, we generally don't care about dev dependencies for those.

The rationale behind this check is that if a package has no `dependencies`, it doesn't need a lockfile symlink next to it, therefore if there _is_ a symlink here, it is a superfluous symlink and should be removed.